### PR TITLE
nixos/stage-1-systemd: Handover between the systemds directly

### LINKED ISSUE #165649: Proper handover between systemds
This is part of the push to switch NixOS's stage 1 to systemd, see also [the project](https://github.com/NixOS/nixpkgs/projects/51).

## Issue description

We are currently switching from the stage 1 systemd to `stage-2-init.sh` (`$out/init`) which does some preparation and starts stage-2 systemd in the end.
The goal here is to not call the script as PID 1 but rather as a service to make it perform the preparations.
If the switchover to stage 2 systemd is performed by the stage 1 systemd directly, the state gets serialized and transferred between the systemds.
This should move the journal and also allows for persistent daemons between the two stages (for example iSCSI).

My current approach is to provide a second script that does everything that is still necessary (there will still be stuff to do like setting up /etc or /run/*-system) and call it in a chroot from stage 1 before switching to the second systemd instance.

#165453 was created to set the environment of stage 2 systemd without stage-2-init.sh which is a first step torwards the right direction.

### PR DESCRIPTION
**Pull request by dasJ**
###### Description of changes

Closes #165649.

Draft because it depends on #167153.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/boot/stage-2-init.sh [hunk 1]
```diff
@@ -47,6 +55,9 @@ fi
 
 
 echo "booting system configuration $systemConfig" > /dev/kmsg
+if [ "${IN_NIXOS_SYSTEMD_STAGE1:-}" = 1 ]; then
```
Comment (arianvp) [inline nixos/modules/system/boot/stage-2-init.sh:58]: if else?

File: nixos/modules/system/boot/systemd/initrd.nix [hunk 2]
```diff
@@ -412,6 +412,60 @@ in {
...
+
+        script = ''
+          set -uo pipefail
+          export PATH="/bin:${cfg.package.util-linux}/bin"
+
+          # Figure out what closure to boot
+          closure=
+          for o in $(< /proc/cmdline); do
+              case $o in
+                  init=*)
```
[the remark this answers is not in the document]
Comment (ElvishJerricco) [reply nixos/modules/system/boot/systemd/initrd.nix:430]: Yea, I'd like for the `init=` cmdline parameter to work correctly... Could we have `init=/run/current-system/systemd/lib/systemd/systemd` somehow?

[the remark this answers is not in the document]
Comment (dasJ) [reply nixos/modules/system/boot/systemd/initrd.nix:430]: Yeah but then we'd need a second parameter to tell us which closure to boot :(

File: nixos/modules/system/boot/systemd/initrd.nix [hunk 3]
```diff
@@ -367,7 +368,7 @@ in {
 
         "/etc/systemd/system.conf".text = ''
           [Manager]
-          DefaultEnvironment=PATH=/bin:/sbin
+          DefaultEnvironment=PATH=/bin:/sbin ${optionalString (isBool cfg.emergencyAccess && cfg.emergencyAccess) "SYSTEMD_SULOGIN_FORCE=1"}
```
Comment (ElvishJerricco) [inline nixos/modules/system/boot/systemd/initrd.nix:371]: Does this cause stage 2 init to have this environment variable set as well?

Comment (dasJ) [reply nixos/modules/system/boot/systemd/initrd.nix:371]: Does not seem to be the case, my services don't have the variable set

**Pull request merged**